File: api/Dtos/State.cs
using api.Utility.Paging;
using System.ComponentModel.DataAnnotations;

namespace api.Dtos
{
    public class StateRes
    {
        public int StateId { get; set; }
        public string? Name { get; set; }
        public string? Code { get; set; }
        public int CountryId { get; set; }
    }

    public class StateResWithCitiesCount : StateRes
    {
        public int CitiesCount { get; set; }
    }

    public class StateResWithCountryAndCitiesCount : StateRes
    {
        public string? CountryName { get; set; }
        public int CitiesCount { get; set; }
    }

    public class StateReqEdit
    {
        [Required, MaxLength(255)]
        public string? Name { get; set; }
        [Required, MaxLength(50)]
        public string? Code { get; set; }

        // Foreign keys
        public int CountryId { get; set; }
    }

    public class StateReqSearch : PagedRequestDto
    {
        public int CountryId { get; set; }
    }
}

File: web/src/dtos/State.ts
import Common from "../utility/Common";
import { PagedReq } from "./PagedReq";

export interface StateRes {
  stateId?: number;
  name?: string;
  code?: string;
  countryId?: number;
}

export interface StateResWithCitiesCount extends StateRes {
  citiesCount?: number;
}

export class StateReqEdit {
  name?: string = "";
  code?: string = "";
  countryId?: number = 0;
}

export class StateReqSearch extends PagedReq {
  countryId?: number;

  constructor(
    {
      pageNumber = 1,
      pageSize = Common.DEFAULT_PAGE_SIZE,
      orderBy,
      searchText = "",
    }: PagedReq,
    countryId = 0
  ) {
    super({
      pageNumber: pageNumber,
      pageSize: pageSize,
      orderBy: orderBy,
      searchText: searchText,
    });
    this.countryId = countryId;
  }
}
